### PR TITLE
keras logger

### LINKED ISSUE #597: tensorboard logging model parameters or training stats for keras models

### PR DESCRIPTION
**Pull request by hstojic**
this closes #597 - logging keras model losses and metrics to the tensorboard, only summary stats available,
epoch loss traces have to be done differently, by providing a callback in  `fit_args`, the only issue is that errors is raised when trying to save the model with tensorboard callback... (see #599

### CODE REVIEW

File: trieste/ask_tell_optimization.py [hunk 1]
```diff
@@ -294,6 +294,13 @@ def __init__(
                 model.update(dataset)
                 model.optimize(dataset)
 
+            # summary_writer = logging.get_tensorboard_writer()
+            # if summary_writer:
+            #     with summary_writer.as_default(step=-1):
+            #         for tag, model in self._models.items():
+            #             with tf.name_scope(f"{tag}.model"):
+            #                 model.log()
+
```
Comment (hstojic) [inline trieste/ask_tell_optimization.py]: @uri-granta we forgot to log model at initial fit, this code is not working, what would be the best way to do this?

File: trieste/bayesian_optimizer.py [hunk 2]
```diff
@@ -675,6 +675,13 @@ def optimize(
...
                                 model.optimize(dataset)
 
+                            # summary_writer = logging.get_tensorboard_writer()
+                            # if summary_writer:
+                            #     with summary_writer.as_default(step=-1):
+                            #         for tag, model in models.items():
+                            #             with tf.name_scope(f"{tag}.model"):
+                            #                 model.log()
+
                     with Timer() as query_point_generation_timer:
```
Comment (hstojic) [inline trieste/bayesian_optimizer.py]: @uri-granta same here?

File: trieste/models/keras/interface.py [hunk 3]
```diff
@@ -73,6 +74,23 @@ def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
...
+        if summary_writer:
+            with summary_writer.as_default(step=logging.get_step_number()):
+                logging.scalar(f"num_epochs", len(self.model.history.epoch))
+                for k, v in self.model.history.history.items():
+                    logging.histogram(f"epoch_{k}", v)
+                    logging.scalar(f"{k}_final", v[-1])
+                    logging.scalar(f"{k}_diff", v[0] - v[-1])
+                    logging.scalar(f"{k}_min", tf.reduce_min(v))
+                    logging.scalar(f"{k}_max", tf.reduce_max(v))
+
```
Comment (hstojic) [inline trieste/models/keras/interface.py]: @willcowley what do you think of these summaries, would you include anything else?

File: trieste/ask_tell_optimization.py (hunk 1 above)
Comment (uri-granta) [reply trieste/ask_tell_optimization.py]: If the issue is `step=-1` then the cleanest solution is probably to use `step=0` here and update the BO loop to start counting from 1 (`for step in range(1, num_steps + 1)`, and remembering to update the `fix_initial_model` check). This would arguably be more user friendly anyway (the first step is currently "step #0") and the only thing apart from logging that depends on the number is the tracked state filenames, which will work fine too.

File: trieste/models/keras/models.py [hunk 4]
```diff
@@ -359,15 +367,82 @@ def optimize(self, dataset: Dataset) -> None:
...
+                        logging.histogram(
+                            f"variance_predict_{i}", predict_ensemble_variance[i, ...]
+                        )
+                        logging.scalar(
+                            f"variance_predict_mean_{i}",
+                            tf.reduce_mean(predict_ensemble_variance[i, ...]),
+                        )
+                    # data stats
+                    empirical_variance = tf.math.reduce_variance(dataset.observations)
+                    logging.scalar("variance_empirical", empirical_variance)
```
Comment (hstojic) [inline trieste/models/keras/models.py]: @willcowley what do you think of the summaries above, would you include anything else?

File: trieste/models/keras/models.py [hunk 5]
```diff
@@ -359,15 +367,82 @@ def optimize(self, dataset: Dataset) -> None:
...
+
+        :param dataset: Optional data that can be used to log additional data-based model summaries.
+        """
+        summary_writer = logging.get_tensorboard_writer()
+        if summary_writer:
+            with summary_writer.as_default(step=logging.get_step_number()):
+                logging.scalar("num_epochs", len(self.model.history.epoch))
+                for k, v in self.model.history.history.items():
+                    logging.histogram(f"{k}_epoch", v)
+                    logging.scalar(f"{k}_final", v[-1])
```
Comment (uri-granta) [inline trieste/models/keras/models.py]: (If there's any chance that these can throw an exception (e.g. v is empty) then you can wrap the logged data in a closure, which will catch the exception.)

```suggestion
                    logging.scalar(f"{k}_final", lambda: v[-1])
```

File: trieste/models/keras/models.py [hunk 6]
```diff
@@ -359,15 +367,82 @@ def optimize(self, dataset: Dataset) -> None:
...
+        For custom logs user will need to subclass the model and overwrite this method.
+
+        :param dataset: Optional data that can be used to log additional data-based model summaries.
+        """
+        summary_writer = logging.get_tensorboard_writer()
+        if summary_writer:
+            with summary_writer.as_default(step=logging.get_step_number()):
+                logging.scalar("num_epochs", len(self.model.history.epoch))
+                for k, v in self.model.history.history.items():
+                    logging.histogram(f"{k}_epoch", v)
```
Comment (uri-granta) [inline trieste/models/keras/models.py]: Perhaps make all these collapsible under `k`? Also what sort of names do `k`s have? Should there be a prefix?

```suggestion
                    logging.histogram(f"{k}/epoch", v)
```

File: trieste/models/keras/models.py [hunk 7]
```diff
@@ -359,15 +367,82 @@ def optimize(self, dataset: Dataset) -> None:
...
+                    logging.scalar(f"{k}_final", v[-1])
+                    logging.scalar(f"{k}_diff", v[0] - v[-1])
+                    logging.scalar(f"{k}_min", tf.reduce_min(v))
+                    logging.scalar(f"{k}_max", tf.reduce_max(v))
+                if dataset:
+                    predict = self.predict(dataset.query_points)
+                    # training accuracy
+                    diffs = predict[0] - tf.cast(dataset.observations, predict[0].dtype)
+                    logging.histogram("accuracy_absolute_errors", tf.math.abs(diffs))
```
Comment (uri-granta) [inline trieste/models/keras/models.py]: Again, I'd collate all `accuracy` logs together, and similarly for `variance` logs.

```suggestion
                    logging.histogram("accuracy/absolute_errors", tf.math.abs(diffs))
```

**Review by uri-granta — APPROVED** [review]
LGTM, though it'd be good to run a screenshot of the logs past someone with a better sense of how useful they are.

File: trieste/models/keras/models.py (hunk 7 above)
Comment (avullo) [reply trieste/models/keras/models.py]: How about logging the histogram of the normalised residuals? You'd get a better indication of the behaviour of the model independent on the scale of the differences.
On a similar note, you could log the Z-score of the normalised residuals (should be close to 1).

File: trieste/models/keras/models.py (hunk 6 above)
Comment (hstojic) [reply trieste/models/keras/models.py]: tried to improve that part further

File: trieste/models/keras/models.py (hunk 7 above)
Comment (hstojic) [reply trieste/models/keras/models.py]: good idea, I have added 
- z residuals histogram
- variance error histogram
- z residuals std scalar
- root mean variance error scalar

File: trieste/models/keras/models.py [hunk 8]
```diff
@@ -409,39 +409,59 @@ def log(self, dataset: Optional[Dataset] = None) -> None:
...
-                logging.scalar("num_epochs", len(self.model.history.epoch))
+                logging.scalar("epochs/num_epochs", len(self.model.history.epoch))
                 for k, v in self.model.history.history.items():
-                    logging.histogram(f"{k}_epoch", v)
-                    logging.scalar(f"{k}_final", v[-1])
-                    logging.scalar(f"{k}_diff", v[0] - v[-1])
-                    logging.scalar(f"{k}_min", tf.reduce_min(v))
-                    logging.scalar(f"{k}_max", tf.reduce_max(v))
+                    k_sep = k.split("_")
+                    if "loss" in k and "model" in k:
```
Comment (uri-granta) [inline trieste/models/keras/models.py]: This is very arcane! Without knowing the contents of history I can't tell what the names will be (and whether this will ever throw an exception due to k_sep being size 1). Maybe a few comments with examples?

**Review by uri-granta — APPROVED** [review]
Still looks good!

File: trieste/models/keras/models.py (hunk 8 above)
Comment (hstojic) [reply trieste/models/keras/models.py]: I was unhappy with this bit, feels a bit brittle to rely on output names assigned by KerasEnsemble - here I was trying to separate loss as a group and identify if any other metrics is in the output (here we have `mse` by default in `DeepEnsemble`) and if so make it a grouping

the alternative is to go with the unprocessed variant but then loss and whatever other metrics are defined will be mixed in ... perhaps I can get metrics names somehow from the compiled model, that would make it cleaner

Comment (hstojic) [reply trieste/models/keras/models.py]: the default output looks like: 
```
{"loss":[...], "model_i_output_loss":[...], "model_i_output_mse":[...]}
```
where i is the member of ensemble, so we would have e.g. 10 of those

Comment (uri-granta) [reply trieste/models/keras/models.py]: Maybe use a regex based whitelist instead then? Clearer and more resilient.

```python
KEY_SPLITTER = {
    # map history keys to prefix and suffix
    "loss" : ("loss", ""),
    r"(?P<model>model_\d+)_output_loss" : ("loss", r"_\g<model>"),
    r"(?P<model>model_\d+)_(?P<metric>.+)" : (r"\g<metric>", r"_\g<model>"),
}
for pattern, (pre_sub, post_sub) in KEY_SPLITTER.items():
    if re.match(pattern, k):
        pre = re.sub(pattern, pre_sub, k)
        post = re.sub(pattern, post_sub, k)
        break
else:
    # unrecognised history key; ignore
    continue
```
